tools/nimgrep.nim

```diff
@@ -12,64 +12,106 @@ import
 
 const
   Version = "1.6.0"
-  Usage = "nimgrep - Nim Grep Utility Version " & Version & """
+  Usage = "nimgrep - Nim Grep Searching and Replacement Utility Version " &
+  Version & """
+
+  (c) 2012-2020 Andreas Rumpf
 
-  (c) 2012 Andreas Rumpf
 Usage:
-  nimgrep [options] pattern [replacement] (file/directory)*
+* To search:
+  nimgrep [options] PATTERN [(FILE/DIRECTORY)*/-]
+* To replace:
+  nimgrep [options] PATTERN --replace REPLACEMENT (FILE/DIRECTORY)*/-
+* To list file names:
+  nimgrep [options] --filenames [PATTERN] [(FILE/DIRECTORY)*]
+
+Positional arguments, from left to right:
+* PATERN is either Regex (default) or Peg if --peg is specified.
+  PATTERN and REPLACEMENT should be skipped when --stdin is specified.
+* REPLACEMENT supports $1, $# notations for captured groups in PATTERN.
+  Note: --replace mode DOES NOT ask confirmation unless --confirm is specified!
+* Final arguments are a list of paths (FILE/DIRECTORY) or a standalone
+  minus '-' (pipe) or not specified (empty). Note for the empty case: when
+  no FILE/DIRECTORY/- is specified nimgrep DOES NOT read the pipe, but
+  searches files in the current dir instead!
+    -                   read buffer once from stdin: pipe or terminal input;
+                        in --replace mode the result is directed to stdout.
+    (empty)             current directory '.' is assumed
```

Comment:
seems dangerous:

`nimgrep foo --replace bar **/*.js` => when glob `**/*.js` returns nothing by the shell, it'd behave the same as `nimgrep foo --replace bar .` which has very different semantics?

Reply:
yes, that's why replace made a special case: it requires explicit specification of all 3 positional arguments (and that is written right 3 lines below), so in this case nimgrep will report an error.